### DISCUSSION TITLE
Task - Spring Boot - Complete CRUD API - REST Backend - my_org DB

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#17] [Tasks]
# CRUD API with Endpoints for `my_org` DB

Create a Spring Boot application named `spring-boot-my-org-crud-api` (artifactId and groupId are in all lowercase) to connect with the `my_org` DB using **Jdbc** (mysql-connector-j). Create the base package of the project named `com.springboot.my.org.crudapi`.

The details for creating the Spring Boot project are as follows:
- name: `spring-boot-my-org-crud-api`
- group: `com.my.org.`
- artifact: `springbootmyorgcrudapi`
- package: `com.springboot.my.org.crudapi`

Dependencies to include:
- Spring Web
- MySQL Driver
- Spring Boot DevTools (optional)
- Spring Configuration Processor (optional)

    ```xml
    <dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>

		<dependency>
			<groupId>mysql</groupId>
			<artifactId>mysql-connector-java</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
    </dependencies>
    ```

Perform _CRUD_ operations on the `Worker` table using models, DAOs, and repositories. 

Create these sub-packages under the base package:
  - `com.springboot.my.org.crudapi.util`
  - `com.springboot.my.org.crudapi.model`
  - `com.springboot.my.org.crudapi.dao`
  - `com.springboot.my.org.crudapi.repository`
  -  `com.springboot.my.org.crudapi.service`
  - `com.springboot.my.org.crudapi.controller`

**Hint - You can refer to discussion #16 for code snippets to reuse.**

Create the helper classes such as `DatabaseConnection`, `DateTimeUtilities` in the `com.springboot.my.org.crudapi.util` package`.

Create the following **model** classes in `com.springboot.workers.crudapi.controller` package:
- `Worker`
- `Bonus`
- `Title`

Create the following **dao** interfaces in `com.springboot.my.org.crudapi.dao` package which contain abstract methods for CRUD operations on the corresponding tables:
- `WorkerDAO`
- `BonusDAO`
- `TitleDAO`

Create the following **repository** classes in `com.springboot.workers.crudapi.repository` package and implement the abstract methods of their respective DAOs:
- `WorkerRepository`
- `BonusRepository`
- `TitleRepository`

Create the following **service** classes in `com.springboot.workers.crudapi.service` package and annotate them with `@Service` annotation; autowire their respective Repository classes to use their methods:
- `WorkerService`
- `BonusService`
- `TitleService`

Create the following thin **controller** classes in `com.springboot.workers.crudapi.controller` package:
- `HomeController`
- `WorkerController`
- `BonusController`
- `TitleController`

Create the handler methods and the following mappings in the appropriate controllers:

- `/` - GET - Displays a welcome message "Hello visitor!\nVisiting time: ${CURRENT_TIMESTAMP}"
-------------------------------------------------------------------------------------------------
- `/worker/{id}` - **GET** - Displays the details of a `Worker` record. Use the `worker_id` from the path variable to display the worker details. Return the worker record as a JSON response.
- `/worker/all` - **GET** - Displays the details of all the workers in the `Worker` table. Return the List of the workers as a JSON response.
- `/worker/create` - **POST** - Create a new worker record. Use the data from RequestBody(raw JSON) to insert a new worker record. Return `true` if the worker was successfully created, else it returns `false`. Send a _201_ (HttpStatus.CREATED) response code upon successful insertion.
- `/worker/update/{id}` - **PATCH** - Update the `email` of the worker with the same `worker_id` as the one in the path variable to "mfs.akash@gmail.com". Take the new email in the request body. Return `true` if the record was updated successfully, else return `false`.
- `/worker/delete/{id}` - **DELETE** - Delete the record of the worker whose `worker_id` matches the `id` in the path variable. Return `true` if the record was successfully deleted, else return `false`.
-------------------------------------------------------------------------------------------------
- `/worker/bonus/{id}` - **GET** - Displays the bonus details of a worker using the `worker_ref_id` from the path variable. Return the worker record as a JSON response.
- `dept/{dept}/bonuses` - **GET** - Displays the bonus details of all the workers in the department. Return the List of the bonuses.
- `worker/{id}/bonus/new` - **POST** - Give a new bonus to the worker. Use the data from RequestBody(raw JSON) to get the bonus amount. Return `true` if the bonus record was successfully created, else it returns `false`. Send a _201_ (HttpStatus.CREATED) response code upon successful insertion.
- `/worker/{id}/bonus/latest` - **DELETE** - Delete the latest bonus record for the given worker. Return `true` if the record was successfully deleted, else return `false`.
-------------------------------------------------------------------------------------------------
- `/worker/title/{id}` - **GET** - Displays the title details of a worker using the `worker_ref_id` from the path variable. Return the worker record as a JSON response.
- `dept/{dept}/all` - **GET** - Displays the full details of all the workers in the department with their `title`. Return the List of the workers.
- `worker/{id}/promote/{title}` - **POST** - Give a promotion to the worker. Use the data from RequestBody(raw JSON) to get the new title. Return `true` if the title record was successfully created, else it returns `false`. Send a _201_ (HttpStatus.CREATED) response code upon successful insertion.
- `/worker/{id}/demote` - **DELETE** - Delete the latest promotion (title) given to a worker and revert to their old title. Id the worker had no previous titles, return `false`. Return `true` if the record was successfully deleted, else return `false`.
-------------------------------------------------------------------------------------------------
- `worker/all/fullInfo` - **GET** - Displays the full details of all the workers in the department with their `title` and `bonus`. Return the List of the workers.
- `dept/{dept}/all/fullInfo` - **GET** - Displays the full details of all the workers in the department with their `title` and `bonus`. Return the List of the workers.
-------------------------------------------------------------------------------------------------

Use Postman to test the API.

**Comment by RohitN09** [comment]
https://github.com/RohitN09/Spring/tree/master/spring-boot-org-crud-api

**Comment by Gaurav-Joshi-31** [comment]
https://github.com/Gaurav-Joshi-31/springprojects/tree/main/spring-boot-my-org-crud-api

**Comment by abhiraj2k** [comment]
https://github.com/abhiraj2k/spring-projects/tree/master/springbootmyorgcrudapi

**Comment by amruth2211** [comment]
https://github.com/amruth2211/Spring-Projects/tree/main/spring-boot-org-crud-api/src/main/java/com/springboot/my/org/crudapi

**Comment by Ashwesh09** [comment]
[Complete Crud API Submission](https://github.com/Ashwesh09/Web-Projects/tree/main/springbootmyorgcrudapi)

**Comment by Harshita-Joshi** [comment]
https://github.com/Harshita-Joshi/Spring/tree/main/spring-boot-my-org-crud-api

**Comment by rakeshGit419** [comment]
[https://github.com/rakeshGit419/Wiley_Java/tree/main/OneDrive/Desktop/Wiley_java/spring-boot-my-org-crud-api](url)

**Comment by Shubhi27** [comment]
https://github.com/Shubhi27/spring_projects/tree/main/spring-boot-my-org-crud-api

**Comment by SrideviYamala** [comment]
https://github.com/SrideviYamala/spring-projects/tree/master/spring-boot-my-org-crud-api

**Comment by Krishreddy460** [comment]
[completecrud](https://github.com/Krishreddy460/crudcomplete)

**Comment by Sarfinaa** [comment]
[CRUD API ](https://github.com/Sarfinaa/SpringAndSpringBootProjects/tree/master/springBootCrudApi)

**Comment by lalit-ky** [comment]
https://github.com/009lalitkumar/Spring-SpringBoot-Tasks/tree/main/17

**Comment by Prem-3** [comment]
[https://github.com/Prem-3/Task/tree/main/spring-boot-my-org-crud-api](url)
